plugins/logitech-tap/fu-logitech-tap-touch-device.c

```diff
@@ -0,0 +1,913 @@
+	fu_byte_array_append_uint8(st, 0x00);
+	return fu_logitech_tap_touch_device_hid_transfer(self, st, 0, NULL, error);
+}
+
+static gboolean
+fu_logitech_tap_touch_device_write_enable(FuLogitechTapTouchDevice *self,
+					  gboolean in_ap,
+					  gboolean write_ap,
+					  guint32 end,
+					  guint32 checksum,
+					  GError **error)
+{
+	guint8 delay;
+	g_autoptr(FuStructLogitechTapTouch) st = fu_struct_logitech_tap_touch_new();
+
+	if (in_ap) {
+		fu_struct_logitech_tap_touch_set_payload_len(st, 0x03);
+		delay = 100;
+	} else {
+		fu_struct_logitech_tap_touch_set_payload_len(st, 0x0A);
+		delay = 10;
+	}
+	fu_struct_logitech_tap_touch_set_response_len(st, 0x0);
+	fu_byte_array_append_uint8(st, FU_STRUCT_LOGITECH_TAP_TOUCH_HID_CMD_WRITE_ENABLE);
+	fu_byte_array_append_uint8(st, 0x5A);
+	fu_byte_array_append_uint8(st, 0xA5);
+	if (end > 0) {
+		fu_byte_array_append_uint8(st, write_ap ? 0x00 : 0x01);
+		fu_byte_array_append_uint24(st, end, G_BIG_ENDIAN);
+		fu_byte_array_append_uint24(st, checksum, G_BIG_ENDIAN);
+	}
+
+	/* hid report to enable writing */
+	if (!fu_logitech_tap_touch_device_hid_transfer(self, st, 0, NULL, error))
+		return FALSE;
+
+	/* mode switch delay for application/bootloader */
+	fu_device_sleep(FU_DEVICE(self), delay);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_touch_device_check_busy_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuLogitechTapTouchDevice *self = FU_LOGITECH_TAP_TOUCH_DEVICE(device);
+	g_autoptr(GByteArray) buf_res = g_byte_array_new();
+	g_autoptr(FuStructLogitechTapTouch) st = fu_struct_logitech_tap_touch_new();
+
+	/* hid report to query device busy or idle status */
+	fu_struct_logitech_tap_touch_set_payload_len(st, 0x01);
+	fu_struct_logitech_tap_touch_set_response_len(st, 0x01);
+	fu_byte_array_append_uint8(st, FU_STRUCT_LOGITECH_TAP_TOUCH_HID_CMD_GET_SYS_BUSY_STATUS);
+	if (!fu_logitech_tap_touch_device_hid_transfer(self, st, 5, buf_res, error))
+		return FALSE;
+	if (buf_res->data[TAP_TOUCH_HID_RESPONSE_OFFSET] != TAP_TOUCH_SYSTEM_READY) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "device busy check failed, got:  0x%02x, expected: %i",
+			    buf_res->data[TAP_TOUCH_HID_RESPONSE_OFFSET],
+			    TAP_TOUCH_SYSTEM_READY);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_touch_device_get_crc(FuLogitechTapTouchDevice *self,
+				     guint16 *crc,
+				     guint8 datasz,
+				     GError **error)
+{
+	g_autoptr(FuStructLogitechTapTouch) st = fu_struct_logitech_tap_touch_new();
+	fu_struct_logitech_tap_touch_set_payload_len(st, 0x01);
+
+	if (crc)
+		fu_struct_logitech_tap_touch_set_response_len(st, datasz);
+	else
+		fu_struct_logitech_tap_touch_set_response_len(st, 0x0);
+	fu_byte_array_append_uint8(st, FU_STRUCT_LOGITECH_TAP_TOUCH_HID_CMD_GET_AP_CRC);
+
+	/* hid report to query crc info of dataflash/pflash (DF/AP) block */
+	if (crc) {
```

Comment:
`!= NULL` please -- then it's obvious it's a pointer and not being compared against non-zero

Reply:
Fixed